src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -0,0 +1,175 @@
+            string[] scopes = new string[] { scope };
+
+            if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryServicePrincipal)
+            {
+                IConfidentialClientApplication ccApp = ConfidentialClientApplicationBuilder.Create(parameters.UserId)
+                    .WithAuthority(parameters.Authority)
+                    .WithClientSecret(parameters.Password)
+                    .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                    .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                    .Build();
+
+                result = ccApp.AcquireTokenForClient(scopes).ExecuteAsync().Result;
+                return new SqlAuthenticationToken(result.AccessToken, result.ExpiresOn);
+            }
+
+            IPublicClientApplication app = PublicClientApplicationBuilder.Create(ActiveDirectoryAuthentication.AdoClientId)
+                .WithAuthority(parameters.Authority)
+                .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                .WithRedirectUri("http://localhost")
+                .Build();
+
+            if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryIntegrated)
+            {
+                result = app.AcquireTokenByIntegratedWindowsAuth(scopes)
+                    .WithCorrelationId(parameters.ConnectionId)
+                    .ExecuteAsync().Result;
+            }
+            else if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryPassword)
+            {
+                SecureString password = new SecureString();
+                foreach (char c in parameters.Password)
+                    password.AppendChar(c);
+                password.MakeReadOnly();
+                result = app.AcquireTokenByUsernamePassword(scopes, parameters.UserId, password)
+                    .WithCorrelationId(parameters.ConnectionId)
+                    .ExecuteAsync().Result;
+            }
+            else if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryInteractive)
+            {
+                var accounts = await app.GetAccountsAsync();
+                IAccount account;
+                if (!string.IsNullOrEmpty(parameters.UserId))
+                {
+                    account = accounts.FirstOrDefault(a => parameters.UserId.Equals(a.Username, System.StringComparison.InvariantCultureIgnoreCase));
+                }
+                else
+                {
+                    account = accounts.FirstOrDefault();
+                }
+
+                if (null != account)
+                {
+                    try
+                    {
+                        result = await app.AcquireTokenSilent(scopes, account).ExecuteAsync();
```

Comment:
This is the Interactive pattern recommended by MSAL to avoid prompts every time. If the user has already gone through it once, this takes advantage of the MSAL token cache.